packages/frontend/src/utils/wallet.js

```diff
@@ -273,6 +282,89 @@ class Wallet {
         })));
     }
 
+    async getPublicKeyType(accountId, publicKeyString) {
+        const allKeys = await this.getAccessKeys(accountId);
+        const keyInfoView = allKeys.find(({public_key}) => public_key === publicKeyString);
+
+        if (keyInfoView) {
+           if (this.isFullAccessKeyInfoView(keyInfoView)) return Wallet.KEY_TYPES.FAK;
+           if (this.isLedgerKeyInfoView(accountId, keyInfoView)) return Wallet.KEY_TYPES.LEDGER;
+           if (this.isMultisigKeyInfoView(accountId, keyInfoView)) return Wallet.KEY_TYPES.MULTISIG;
+            return Wallet.KEY_TYPES.OTHER;
+        }
+
+        throw new Error('No matching key pair for public key');
+    }
+
+    isFullAccessKeyInfoView(keyInfoView) {
+        return keyInfoView?.access_key?.permission === 'FullAccess';
+    }
+
+    isLedgerKeyInfoView(accountId, keyInfoView) {
+        const receiver_id = keyInfoView?.access_key?.permission?.FunctionCall?.receiver_id;
+        const method_names = keyInfoView?.access_key?.permission?.FunctionCall?.method_names;
+        return receiver_id === accountId && isEqual(method_names, ['__wallet__metadata']);
+    }
+
+    isMultisigKeyInfoView(accountId, keyInfoView) {
+        const receiver_id = keyInfoView?.access_key?.permission?.FunctionCall?.receiver_id;
+        const method_names = keyInfoView?.access_key?.permission?.FunctionCall?.method_names;
+        return receiver_id === accountId && isEqual(method_names, ['add_request', 'add_request_and_confirm', 'delete_request', 'confirm']);
+    }
+
+    async addExistingAccountKeyToWalletKeyStore(accountId, keyPair, ledgerHdPath) {
+        const keyType = await this.getPublicKeyType(
+            accountId,
+            keyPair.getPublicKey().toString()
+        );
+
+        switch (keyType) {
+            case Wallet.KEY_TYPES.FAK: {
+                const keyStore = new nearApiJs.keyStores.InMemoryKeyStore();
+                await keyStore.setKey(NETWORK_ID, accountId, keyPair);
+                const newKeyPair = nearApiJs.KeyPair.fromRandom('ed25519');
+                const account = new nearApiJs.Account(
+                    nearApiJs.Connection.fromConfig({
+                        networkId: NETWORK_ID,
+                        provider: {
+                            type: 'JsonRpcProvider',
+                            args: { url: NODE_URL + '/' },
+                        },
+                        signer: new nearApiJs.InMemorySigner(keyStore),
+                    }),
+                    accountId
+                );
+
+                await account.addKey(newKeyPair.getPublicKey());
+                await this.saveAccount(accountId, newKeyPair);
+                
+                if (!this.accountId) {
+                    return this.makeAccountActive(accountId);
+                }
+                return this.save();
```

Comment:
I dont think this needs to be that DRY. In this case I lean more towards making sure the logic for each `KEY_TYPE` is encapsulated since how they are handled wrt to localStorage is seriously hard to figure out in our code base 😆 . Refactor due soon hopefully with something like redux-persist

Reply:
It's not about being DRY just for the sake it. Making it clear that _all_ paths in this method change the wallet class state in the same way, completely independently of some value derived from the input, is pretty important to how this method behaves. I don't see how localstorage or redux-persist has any bearing on this.

There's really no reason to duplicate this logic across three `case` statements, and it becomes more confusing because now at first glance you would assume there's some subtle difference in how `keyType` changes the behavior of this method.

Reply:
Will possibly iterate on this further in a follow up PR. Maybe 3 if blocks with common logic after work but then I think we may have to throw conditionally. I'll merge this in first though to unblock testing. Thanks for the suggestions 👍